docs/references/http-gateway-protocol-spec.md

```diff
@@ -0,0 +1,471 @@
+STRING-LIST = '[', { STRING }, ']'
+
+RESPONSE-HEADER-LIST = 'ResponseHeaderList{headers:',  STRING-LIST, '}'
+
+REQUEST-CERTIFICATION = 'RequestCertification{certified_request_headers:', STRING-LIST, ',certified_query_parameters:', STRING-LIST, '}'
+
+RESPONSE-CERTIFICATION = 'ResponseCertification{', ('response_header_exclusions:' | 'certified_response_headers:'), RESPONSE-HEADER-LIST, '}'
+
+CERTIFICATION = 'Certification{('no_request_certification: Empty{}' | 'request_certification:', REQUEST-CERTIFICATION), ',response_certification:', RESPONSE-CERTIFICATION, '}'
+
+VALIDATION-ARGS = 'ValidationArgs{', ('no_certification: Empty{}' | 'certification:', CERTIFICATION), '}'
+
+HEADER-VALUE = 'default_certification(', VALIDATION-ARGS, ')'
+
+HEADER = 'IC-CertificateExpression: ', HEADER-VALUE
+```
+
+### Request Hash Calculation
+
+The request hash is calculated as follows:
+
+1. Let `request_headers_hash` be the [representation-independent hash](https://internetcomputer.org/docs/current/references/ic-interface-spec#hash-of-map) of the request headers:
+   - Only include headers listed in the `certified_request_headers` field of [the certificate expression header](#the-certificate-expression-header).
+     - If the field is empty or no value was supplied, no headers are included.
+     - Headers can be repeated and each repetition should be included.
+   - Include an additional `:ic-cert-method` header that contains the HTTP method of the request.
+   - Include an additional `:ic-cert-query` header that contains a value according to the following steps:
+     - Parse the query string and build a list of tuples `(<query_param_name>, <query_param_value>)` while maintaining the order.
+     - Exclude all tuples where `<query_param_name>` does not exactly match a value listed in the `certified_query_parameters` field of [the certificate expression header](#the-certificate-expression-header). If `certified_query_parameters` is empty then the resulting list of tuples should also be empty.
+     - Concatenate each `<query_param_name>` with the corresponding `<query_param_value>` and then concatenate all of these concatenations using the original separators and order.
+     - Calculate the sha256 hash of the UTF-8 representation of the resulting string.
+2. Let `request_body_hash` be the sha256 of the request body.
+3. Concatenate `request_headers_hash` and `request_body_hash` and calculate the sha256 of that concatenation.
+
+### Response Hash Calculation
+
+The response hash is calculated as follows:
+
+1. Let `response_headers_hash` be the [representation-independent hash](https://internetcomputer.org/docs/current/references/ic-interface-spec#hash-of-map) of the response headers:
+   - The `IC-Certificate` header is always excluded.
+   - The `IC-CertificateExpression` header is always included.
+   - ## If the `certified_response_headers` field of [the certificate expression header](#the-certificate-expression-header) is present:
```

Comment:
```suggestion
   - If the `certified_response_headers` field of [the certificate expression header](#the-certificate-expression-header) is present:
```

Reply:
You're right... I'll fill this in, thanks!